### PR TITLE
feat(user-segments): segments management APIs

### PR DESCRIPTION
**Pull request by jingnanl**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

User segments APIs

## Implementation highlights

(describe how the merge request does for feature changes, share the RFC link if it has)

## Test checklist

- [ ] add new test cases
- [ ] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [ ] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [ ] deploy data modeling
    - [ ] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting
  - [ ] streaming ingestion
    - [ ] with Redshift Serverless
    - [ ] with provisioned Redshift

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

File: src/control-plane/backend/click-stream-api.ts [hunk 1]
```diff
@@ -406,6 +412,15 @@ export class ClickStreamApiConstruct extends Construct {
               arnFormat: ArnFormat.COLON_RESOURCE_NAME,
             }, Stack.of(this),
           ),
+          Arn.format(
+            {
+              resource: 'stateMachine',
+              service: 'states',
+              region: '*',
+              resourceName: `${CLICKSTREAM_SEGMENTS_WORKFLOW_PREFIX}*`,
```
Comment (zxkane) [inline src/control-plane/backend/click-stream-api.ts:420]: You can merge this resource into the above policies.

File: src/control-plane/backend/lambda/api/service/segment.ts [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+const eventBridgeClient = new EventBridgeClient({
+  ...aws_sdk_client_common_config,
+  region: process.env.AWS_REGION,
+});
+const sfnClient = new SFNClient({
+  ...aws_sdk_client_common_config,
+  region: process.env.AWS_REGION,
+});
+const s3Client = new S3Client({
+  ...aws_sdk_client_common_config,
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/service/segment.ts:54]: should be the pipeline region?

File: src/control-plane/backend/lambda/api/service/segment.ts [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+      return res.status(201).json(new ApiSuccess({ id }, 'Segment created successfully.'));
+    } catch (error) {
+      return next(error);
+    }
+  }
+
+  public async list(req: Request, res: Response, next: NextFunction) {
+    try {
+      const { appId } = req.query;
+      const segments = await segmentStore.list(appId as string);
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/service/segment.ts:84]: Do we support pagination?

File: src/control-plane/backend/lambda/api/service/segment.ts [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+      return res.status(200).send(new ApiSuccess({ segmentId }, 'Segment deleted successfully.'));
+    } catch (error) {
+      return next(error);
+    }
+  }
+
+  public async listJobs(req: Request, res: Response, next: NextFunction) {
+    try {
+      const { segmentId } = req.params;
+      const segmentJobs = await segmentStore.listJobs(segmentId as string, SEGMENT_JOBS_LIST_LIMIT);
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/service/segment.ts:177]: Do we support pagination?

File: src/control-plane/backend/lambda/api/service/segment.ts (hunk 2 above)
Comment (zxkane) [reply src/control-plane/backend/lambda/api/service/segment.ts:54]: We have to defer initiating this client based on the region of the pipeline.

File: src/control-plane/backend/lambda/api/service/segment.ts [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+import { DynamoDBSegmentStore } from '../store/dynamodb/dynamodb-segment-store';
+import { DynamoDbStore } from '../store/dynamodb/dynamodb-store';
+
+const segmentStore = new DynamoDBSegmentStore();
+const pipelineStore = new DynamoDbStore();
+const eventBridgeClient = new EventBridgeClient({
+  ...aws_sdk_client_common_config,
+  region: process.env.AWS_REGION,
+});
+const sfnClient = new SFNClient({
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/service/segment.ts:49]: I remember it's a resource in the pipeline stack; we should defer initiating this client based on the region of the given pipeline.

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/control-plane/backend/click-stream-api.ts (hunk 1 above)
Comment (jingnanl) [reply src/control-plane/backend/click-stream-api.ts:420]: I'm appending new resource to existing resources list. Is there other place I can merge it to?

File: src/control-plane/backend/lambda/api/service/segment.ts (hunk 2 above)
Comment (jingnanl) [reply src/control-plane/backend/lambda/api/service/segment.ts:47]: Move SFNClient and EventBridgeClient initiation to APIs.

File: src/control-plane/backend/lambda/api/service/segment.ts (hunk 3 above)
Comment (jingnanl) [reply src/control-plane/backend/lambda/api/service/segment.ts:77]: We don't support pagination for segment list yet. Can we assume that user won't have too many segments (like more than 100) for one single app?

File: src/control-plane/backend/lambda/api/service/segment.ts (hunk 4 above)
Comment (jingnanl) [reply src/control-plane/backend/lambda/api/service/segment.ts:170]: For segment jobs, we'll get latest 30 (by default) segment jobs. No pagination is needed.

File: src/control-plane/backend/lambda/api/service/segment.ts (hunk 5 above)
Comment (jingnanl) [reply src/control-plane/backend/lambda/api/service/segment.ts]: Move SFNClient and EventBridgeClient initiation to APIs.

File: src/control-plane/backend/lambda/api/service/segment.ts (hunk 3 above)
Comment (zxkane) [reply src/control-plane/backend/lambda/api/service/segment.ts:77]: That's fine now.

File: src/control-plane/backend/lambda/api/service/segment.ts (hunk 2 above)
Comment (zxkane) [reply src/control-plane/backend/lambda/api/service/segment.ts:47]: Does it work if the pipeline and S3 bucket in the opt-in regions?

Comment (jingnanl) [reply src/control-plane/backend/lambda/api/service/segment.ts]: Move s3Client initiation to API as well.

**Review by zxkane — APPROVED** [review]

**Review by zxkane — APPROVED** [review]

**Review by tyyzqmf — APPROVED** [review]

**Pull request merged**